### PR TITLE
Feature: search diagram panel

### LINKED ISSUE #279: Project page (index page): add diagrams search panel
## Description

On the project page (index page), add a search panel above the diagrams list

To implements:
- search bar to filter the diagram list by name
- multi select with autocomplete filter for all plugin tag. A diagram is linked to a plugins that have tag. If we select a tag, it display only diagrams that linked this tag.

Not to implements:
  - "filters" with "+" buttons
  - environment select

:warning: We don't have environment select and "Filters +" but put the input search at the right of the panel, and the tag select juste before it (so on the left of input search, on the rigth in the panel), not sure If I'm clear or not :disappointed: 

## Mockup

image

### PR DESCRIPTION
**Pull request by Zorin95670**
Resolved #279

### CODE REVIEW

File: cypress/support/step_definitions/event.js [hunk 1]
```diff
@@ -36,12 +36,12 @@ When('I drag {string} onto {string}', (templateOriginSelector, templateDestinati
...
   cy.get(selector)
-    .click();
-  cy.get(option)
-    .click();
-  cy.get(selector)
-    .click();
+    .click()
+    .get(option)
+    .click()
+    .wait(500);
```
Comment (justinehell) [inline cypress/support/step_definitions/event.js:44]: try without disabling eslint
(see `cypress/support/step.../html.js line 80-83)

File: src/components/inputs/TagSelectInput.vue [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      >
+        {{ tag }}
+      </q-chip>
+    </template>
+    <template #option="{ itemProps, opt, selected, toggleOption }">
+      <q-item v-bind="itemProps">
+        <q-item-section side>
+          <q-checkbox
+            color="accent"
+            :model-value="selected"
```
Comment (justinehell) [inline src/components/inputs/TagSelectInput.vue:39]: can you put `:model-value` first please ?

**Review by justinehell — CHANGES_REQUESTED** [review]

File: changelog.md [hunk 3]
```diff
@@ -31,6 +31,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
   * Add tags list to filter project.
   * Add search bar for projects.
   * Add search bar for templates.
+* Index page:
+  * Add a search bar to filter diagrams
```
Comment (justinehell) [inline changelog.md:35]: missing `.`

**Review by justinehell — CHANGES_REQUESTED** [review]

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    Then I expect current url is '{{projectName}}/models'
+
+    # Model3 creation
+    When I click on '[data-cy="create-model-button"]'
+    And  I set on '[data-cy="create-model-form"] [data-cy="name-input"]' text '{{modelName3}}'
+    And  I select '[data-cy="item_githubator-plugin"]' in '[data-cy="create-model-form"] [data-cy="plugin-select"]'
+    And  I click on '[data-cy="create-model-form"] [data-cy="submit-button"]'
+    Then I expect current url is '{{projectName}}/modelizer/draw\?path=githubator-plugin/{{modelName3}}'
+
+    # Return to project page and verify all models exists
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:44]: all diagrams exist

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 5]
```diff
@@ -0,0 +1,87 @@
+Feature: Test home page: diagram filter
+
+  Scenario: Should filter project by tag and search text
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:3]: Should filter diagrams by tags

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    Then I expect current url is '{{projectName}}/modelizer/draw\?path=githubator-plugin/{{modelName3}}'
+
+    # Return to project page and verify all models exists
+    When I click on '[data-cy="models-page-link-button"]'
+    Then I expect current url is '{{projectName}}/models'
+    And  I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName1}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName2}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/githubator-plugin/{{modelName3}}"]' exists
+
+    # Filter by text and verify that some project disappear
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:51]: some diagrams disappear

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    When I set on '[data-cy="search-diagram-input"]' text '1'
+    Then I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName1}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName2}}"]' not exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/githubator-plugin/{{modelName3}}"]' not exists
+
+    # Clear filter text and verify all models are present
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:57]: all diagrams are

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    When I click on '[data-cy="create-model-button"]'
+    And  I set on '[data-cy="create-model-form"] [data-cy="name-input"]' text '{{modelName2}}'
+    And  I click on '[data-cy="create-model-form"] [data-cy="submit-button"]'
+    Then I expect current url is '{{projectName}}/modelizer/draw\?path=terrator-plugin/{{modelName2}}'
+
+    # Return to project page
+    When I click on '[data-cy="models-page-link-button"]'
+    Then I expect current url is '{{projectName}}/models'
+
+    # Model3 creation
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:37]: diagram 3

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    When I click on '[data-cy="create-model-button"]'
+    And  I set on '[data-cy="create-model-form"] [data-cy="name-input"]' text '{{modelName1}}'
+    And  I click on '[data-cy="create-model-form"] [data-cy="submit-button"]'
+    Then I expect current url is '{{projectName}}/modelizer/draw\?path=terrator-plugin/{{modelName1}}'
+
+    # Return to project page
+    When I click on '[data-cy="models-page-link-button"]'
+    Then I expect current url is '{{projectName}}/models'
+
+    # Model2 creation
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:27]: diagram 2

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    And   I set context field 'modelName1' with 'modelTerrator1'
+    And   I set context field 'modelName2' with 'modelTerrator2'
+    And   I set context field 'modelName3' with 'modelGithubator3'
+
+    When I click on '[data-cy="create-project-button"]'
+    And  I set on '[data-cy="create-project-form"] [data-cy="name-input"]' text '{{projectName}}'
+    And  I click on '[data-cy="create-project-form"] [data-cy="submit-button"]'
+    Then I expect current url is '{{projectName}}/models'
+
+    # Model1 creation
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:17]: diagram 1

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 11]
```diff
@@ -0,0 +1,87 @@
+Feature: Test home page: diagram filter
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:1]: Test models page

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    And  I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName2}}"]' not exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/githubator-plugin/{{modelName3}}"]' not exists
+
+    # Clear filter text and verify all models are present
+    When I set on '[data-cy="search-diagram-input"]' text ' '
+    Then I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName1}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName2}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/githubator-plugin/{{modelName3}}"]' exists
+
+    # Select Terraform tag and verify only all terraform models are present
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:63]: diagrams are present

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    When I select '[data-cy="select-checkbox_Terraform"]' in '[data-cy="diagram-tag-select"]'
+    And  I click on '[data-cy="diagram-tag-select"]'
+    Then I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName1}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName2}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/githubator-plugin/{{modelName3}}"]' not exists
+
+    # Select Github tag and verify all models are present
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:70]: diagrams

File: cypress/e2e/ModelsPage/filter.model.feature [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    When I select '[data-cy="select-checkbox_Github"]' in '[data-cy="diagram-tag-select"]'
+    And  I click on '[data-cy="diagram-tag-select"]'
+    Then I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName1}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/terrator-plugin/{{modelName2}}"]' exists
+    And  I expect '[data-cy="diagram-path_{{projectName}}/githubator-plugin/{{modelName3}}"]' exists
+
+    # Unselect Terraform tag and verify only all githubator models are present
```
Comment (k-y0u) [inline cypress/e2e/ModelsPage/filter.model.feature:77]: diagrams

Comment (k-y0u) [reply cypress/e2e/ModelsPage/filter.model.feature:77]: same for the rest of the file

File: src/components/ModelsView.vue [hunk 15]
```diff
@@ -20,6 +20,13 @@
           })"
         />
       </div>
+      <diagram-filter-card
+        v-model:selected-tags="selectedTags"
```
Comment (k-y0u) [inline src/components/ModelsView.vue:24]: 2 way data binding is real necessary ?
https://vuejs.org/guide/components/v-model.html#multiple-v-model-bindings

**Review by k-y0u — CHANGES_REQUESTED** [review]

File: cypress/support/step_definitions/event.js (hunk 1 above)
Comment (Zorin95670) [reply cypress/support/step_definitions/event.js:44]: Done except for eslint waiting warning

File: src/components/inputs/TagSelectInput.vue (hunk 2 above)
Comment (Zorin95670) [reply src/components/inputs/TagSelectInput.vue:39]: done

File: changelog.md (hunk 3 above)
Comment (Zorin95670) [reply changelog.md:35]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 4 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:44]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 5 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:3]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 7 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:57]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 8 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:37]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 9 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:27]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 10 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:17]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 12 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:63]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 13 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:70]: done

File: cypress/e2e/ModelsPage/filter.model.feature (hunk 14 above)
Comment (Zorin95670) [reply cypress/e2e/ModelsPage/filter.model.feature:77]: done

**Review by k-y0u — APPROVED** [review]

**Review by justinehell — APPROVED** [review]

**Pull request merged**